bundles/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/internal/handler/EnergenieSocket.java

```diff
@@ -0,0 +1,183 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.net.UnknownHostException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.core.util.HexUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class handling the Socket connections.
+ *
+ * @author Hans-Jörg Merk - Initial contribution
+ * @author Hilbrand Bouwkamp - Moved Socket to it's own class
+ */
+@NonNullByDefault
+class EnergenieSocket {
+
+    private static final int SOCKET_TIMEOUT_MILLISECONDS = 1500;
+    private static final byte[] MESSAGE = { 0x11 };
+
+    private final Logger logger = LoggerFactory.getLogger(EnergenieSocket.class);
+    private final String host;
+    private final byte[] key;
+
+    public EnergenieSocket(final String host, final String password) {
+        this.host = host;
+        this.key = getKey(password);
+    }
+
+    private static byte[] getKey(final String password) {
+        final int passwordLength = password.length();
+        String passwordString = password;
+        for (int i = 0; i < (8 - passwordLength); i++) {
+            passwordString = passwordString + " ";
+        }
+        return passwordString.getBytes();
+    }
+
+    public synchronized byte[] sendCommand(final byte[] ctrl) throws IOException {
+        try (final TaskSocket taskSocket = authorize()) {
+            final OutputStream output = taskSocket.socket.getOutputStream();
+            final InputStream input = taskSocket.socket.getInputStream();
+
+            if (output == null || input == null) {
+                throw new IOException("No connection");
+            } else {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Control message send to EG (int) '{}' (hex)'{}'", ctrl, HexUtils.bytesToHex(ctrl));
+                }
+                output.write(encryptControls(ctrl, taskSocket.task));
+                output.flush();
+                readStatus(input, taskSocket);
+                return updateStatus(taskSocket);
+            }
+        }
+    }
+
+    public synchronized byte[] retrieveStatus() throws IOException {
+        try (final TaskSocket taskSocket = authorize()) {
+            return updateStatus(taskSocket);
+        }
+    }
+
+    private TaskSocket authorize() throws IOException {
+        final TaskSocket taskSocket = new TaskSocket();
+        final OutputStream output = taskSocket.socket.getOutputStream();
+        final InputStream input = taskSocket.socket.getInputStream();
+
+        if (output == null || input == null) {
+            throw new IOException("No connection");
+        }
+        output.write(MESSAGE);
+        output.flush();
+        logger.trace("Start Condition '{}' send to EG", MESSAGE);
+        input.read(taskSocket.task);
```

Comment:
This would probably be the easiest approach.
```suggestion
        final DataInputStream input = new DataInputStream(taskSocket.socket.getInputStream());

        if (output == null || input == null) {
            throw new IOException("No connection");
        }
        output.write(MESSAGE);
        output.flush();
        logger.trace("Start Condition '{}' send to EG", MESSAGE);
        input.readFully(taskSocket.task);
```

Reply:
Done